examples/moses-pln-synergy/scm/README.md

```diff
@@ -76,16 +76,204 @@ And search for the following
    )
 ...
 
-scheme@(guile-user)> (cog-bind pln-rule-eliminate-neutral-element-hack)
-scheme@(guile-user)> (cog-bind pln-rule-eliminate-dangling-junctor-hack)
-scheme@(guile-user)> (cog-bind pln-rule-equivalence-hack)
-$6 = (SetLink
+scheme@(guile-user)> (cog-bind implication-lambda-distribution-rule)
+...
    (ImplicationLink (stv 1 0.99999982)
-      (PredicateNode "take-treatment-1" (stv 0.1 0.80000001))
-      (PredicateNode "take-compound-A" (stv 0.2 0.80000001))
+      (LambdaLink
+         (TypedVariableLink
+            (VariableNode "$X")
+            (TypeNode "ConceptNode")
+         )
+         (AndLink
+            (EvaluationLink
+               (PredicateNode "take")
+               (ListLink
+                  (VariableNode "$X")
+                  (ConceptNode "treatment-1")
+               )
+            )
+            (EvaluationLink (stv 1 0.99999982)
+               (PredicateNode "contain")
+               (ListLink
+                  (ConceptNode "treatment-1")
+                  (ConceptNode "compound-A")
+               )
+            )
+         )
+      )
+      (LambdaLink
+         (TypedVariableLink
+            (VariableNode "$X")
+            (TypeNode "ConceptNode")
+         )
```

Comment:
@linas in my view, in Atomese/PLN neither `A` nor `L` nor `M` are correct. At best they are confusing, at worse they are ill-defined.

The only basically correct notations is:

```
ImplicationLink
   P
   Q
```

where P and Q are PredicateNodes or something equivalent to it, like a defined or grounded predicates, or lambdas.

Another notation exists due to its convenience and terseness:

```
ImplicationLink
   <variables>
   <P-body>
   <Q-body>
```

which is equivalent to

```
ImplicationLink
   LamdaLink
      <variables>
      <P-body>
   LambdaLink
      <variables>
      <Q-body>
```

as explained in the wiki page about ImplicationLink.

Can we do more than that and use free variables? I think we can, but for starter it's safer to avoid implicit scoping, that is why I've chosen to clutter the demo with lambdas (ultimately they could be avoided if you use more defined predicates along the way, but I wanted to see how it looks if I manipulate lambda expressions directly).

Now regarding instantiation: please note that there is no universal quantification over ImplicationLink, instead it is condition quantification, P being the condition. When you perform instantiation with an ImplicationLink expression, you don't end up with an instantiated ImplicationLink, you directly end up with an instantiated implicand. Why? Because an instantiated ImplicationLink isn't even defined anyway, and you don't need it, it's more efficient to get what matters, the instantiated implicand.

The ImplicationLink wiki doesn't go enough in depth, I'll try to remedy that soon.